Configapi uma and openId protect code

pujavs opened the pull request:
@yuriyz Request you to please review the code.
- Based on discussion and suggestion from @duttarnab had changed the code from using keystore to using clientId-clientSecret to generate PAT.
- To test the code i had gnerated the token using the oxTrust curl 
curl -k -u '<clientId:clientSecret>' https://pujavs.gluu42.server/oxauth/restv1/token -d grant_type=client_credentials
- However for both OpenId and UMA I am getting "401 Unauthorized" error. Will talk to @duttarnab tomorrow.
- Attached are logs for your reference. 



[oxauth_UMA_API_Protection_17Nov2020_3.logs.pdf](https://github.com/JanssenProject/jans-config-api/files/5555112/oxauth_UMA_API_Protection_17Nov2020_3.logs.pdf)


[oxauth_OPENID_api_protection_17Nov2020_4.logs.pdf](https://github.com/JanssenProject/jans-config-api/files/5555116/oxauth_OPENID_api_protection_17Nov2020_4.logs.pdf)

yuriyz commented on src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java:
```diff
@@ -172,17 +176,20 @@ public void createResourceIfNeeded() {
+    // scope inum + orgnization rather than name
+    // Example: displayName = oxTrust api Resource , oxAuthUmaScope
+    // =inum=1122-BBCC,ou=scopes,o=gluu
+    // Is this fine?
```
We should operate with name. Name is saved under `oxId` attribute of the `Scope` OC.

https://github.com/GluuFederation/oxAuth/blob/fd05fd5ee94cb5553583fadb009a3ecfd1c795bc/persistence-model/src/main/java/org/oxauth/persistence/model/Scope.java#L34

pujavs replied:
Done

yuriyz commented on src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java:
```diff
@@ -172,17 +176,20 @@ public void createResourceIfNeeded() {
             Scope scope = UmaResourceProtectionCache.getScope(strScope);
             if (scope != null) {
-                scopes.add(scope.getInum()+","+scope.getDn());
+                scopes.add(scope.getInum() + "," + scope.getDn());
```
I'm not sure what would that means "inum + dn", if DN is already there then what is the sense to add inum to dn ?

Reverse this logic to work based on scope names (not inum or dn). I you fetch resource from persistence, it references DN, so load scope and take name of it (`oxId`).

pujavs replied:
Like i mentioned in the comments i appended the "inum + dn" by looking at the sample scopes in Gluu schema, thinking that I was doing the right thing keeping it in sync with other gluu code.
However will now rework the logic based on name.

yuriyz replied:
DN = `inum=` + inum + baseDN. While here you seems to get already constructed dn. Even if you tried to construct it it's not correct due to missed `inum=`. Anyway, lets make it via names.

pujavs replied:
Changes done.

yuriyz commented on src/main/java/io/jans/configapi/auth/client/OpenIdClientService.java:
```diff
@@ -1,11 +1,14 @@
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
 
+@RegisterRestClient
+@ApplicationScoped
```
Why do we need duplicate already defined interface from jans-auth-client ? 

We have it here:
https://github.com/JanssenProject/jans-auth-server/blob/56dbfd807233e509a824b23122489d6c266fcce7/client/src/main/java/io/jans/as/client/service/IntrospectionService.java#L22

Is it because quarkus refuse to work with it ? If yes, can you post exact exception or describe problem ?

pujavs replied:
Yes due to resteasy dependency conflict with Quarkus. 
For reference here are the issues of few conflicts - Quarkus_Issue_Resolution.txt.pdf

yuriyz replied:
Nice report. It seems it's not easy to get quarkus work with jans-auth-client . Understood.

yuriyz commented on src/main/java/io/jans/configapi/auth/client/UMATokenService.java:
```diff
@@ -0,0 +1,65 @@
+    @POST
+    @Path("/token")
+    @Produces({MediaType.APPLICATION_JSON})
+    Response requestAccessToken(
```
This is covered by `TokenClient`.

https://github.com/GluuFederation/oxAuth/blob/647bc865d7f121e3da4ab360a59b4b7ed6fd456c/Client/src/main/java/org/gluu/oxauth/client/TokenClient.java#L22

pujavs replied:
Reason is same that is due to resteasy dependency conflict with Quarkus.  The PAT generation code worked by adding this restclient

yuriyz replied:
Clear.

yuriyz commented on src/main/java/io/jans/configapi/auth/model/Condition.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Condition {
```
It's copy-paste from `uma-rs` project. Please add dependency and use library/dependency.

pujavs replied:
I know its copy paste but the uma-rs code was not working due to resteasy conflict and i thought rather than adding the dependency for only these 3 class i would rathermove them within config-api. 
However will add the dependency for these 3 classes if you say so...

yuriyz replied:
Yes, please. Because we may add or remove something in future and then it would lead to redundant work in future.

pujavs replied:
Changes done, now using model classes directly from  jans-client-api

yuriyz reviewed (changes requested):
There many copy-pastes from `uma-rs`, we should re-use code.

pujavs commented:
> There many copy-pastes from `uma-rs`, we should re-use code.
uma-rs could not be reused due to resteasy dependency conflict with Quarkus. Remember earlier when using auth-client also we were getting the conflicts and we decided to code clients within config-api itself.
For reference here are the issues of few conflicts - [Quarkus_Issue_Resolution.txt.pdf](https://github.com/JanssenProject/jans-config-api/files/5558495/Quarkus_Issue_Resolution.txt.pdf)

pujavs commented on src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java:
```diff
@@ -130,8 +132,10 @@ public void createResourceIfNeeded() {
-                String umaResourceName = condition.getHttpMethods() + ":::" + rsResource.getPath(); //??todo: Puja -> To be reviewed by Yuriy Z
-                                                      
+                String umaResourceName = condition.getHttpMethods() + ":::" + rsResource.getPath(); // ??todo: Puja ->
+                                                                                                    // To be reviewed by
```
Done

yuriyz commented:
All these problems makes me think that we should move away from Quarkus in `jans-config-api` (or otherwise move `jans-auth-server` to Quarkus). cc @nynymike

yuriyz commented on src/main/java/io/jans/configapi/auth/UmaResourceProtectionService.java:
```diff
@@ -175,22 +173,11 @@ public void createResourceIfNeeded() {
-            if (scope != null) {
-                scopes.add(scope.getInum() + "," + scope.getDn());
-            }
+            scopes.add(strScope);
```
It's not clear why do we need this method. We have list of scopes and returns same list of scopes. If you remove this method it shouldn't effect anything, right?

pujavs replied:
Correct, its not required any more and hence removed the same.

pujavs commented:
@yuriyz , I modified jans-config code for oauth scope as discussed and have also updated swagger api doc for appropriate scope. 
Request review and merge with master if appraopraite.

**Note**: Have tested oauth for auth properties resource. Attached are the test report and jans-auth-server logs
1) Karate Test Report for properties
[src.test.resources.feature.config.properties.properties.pdf](https://github.com/JanssenProject/jans-config-api/files/5668042/src.test.resources.feature.config.properties.properties.pdf)

2) local jans-auth-server logs
[jans-auth_9Dec2020_OAuth_Final_Working.log](https://github.com/JanssenProject/jans-config-api/files/5668065/jans-auth_9Dec2020_OAuth_Final_Working.log)

yuriyz commented on src/main/java/io/jans/configapi/auth/service/UmaService.java:
```diff
@@ -47,75 +52,67 @@
             RptIntrospectionResponse rptStatusResponse = getStatusResponse(patToken, rptToken);
-            log.trace("RPT status response: {} ", rptStatusResponse);
+            log.debug("RPT status response: {} ", rptStatusResponse);
+            log.debug("RPT status response: {} "+ rptStatusResponse);
```
Duplicate entries, remove it.

yuriyz commented on src/main/java/io/jans/configapi/auth/service/UmaService.java:
```diff
@@ -47,75 +52,67 @@
-                if (rptHasPermissions) {
+                if (rptHasPermissions) {                    
+                    // Verify exact resource
+                    boolean hasResourcePermission = this.hasResourcePermission(rptStatusResponse, resourceId);
```
Now it looks correct.

yuriyz commented:
It's working state but requires improvements in many places. This will be improved in separate PRs by @pujavs
